### ISSUE TITLE
alarms ARE broken in 5.8.0 (FreeBSD 4.6.2)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5929]
Migrated from [rt.perl.org#17341](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=17341) (status was 'resolved')

Searchable as RT17341$

**Comment by p5pRT** [comment]
### From briank@briank.com

The following script demonstrates that the alarm\(\) call is broken
in 5\.8\.0\, at least while running on FreeBSD 4\.6\.2&#8203;:

> cat alarmtest\.pl 
$timeout = 10;
\# use a host that will cause gethostbyname to wait for > $timeout seconds
$badname = "197\.201\.144\.198\.orbs\.gst\-group\.co\.uk";

sub Log\(\) \{
&nbsp;  $| = 1;
&nbsp;  local\($d\) = \`date\`; chop\($d\);
&nbsp;  print "$d @&#8203;\_\\n"
\}

sub myALRM \{ die "alarm\\n" \}
$SIG\{ALRM\} = 'myALRM';

eval \{
&nbsp;  &Log\("alarm\($timeout\)"\);           alarm\($timeout\);
&nbsp;  &Log\("gethostbyname\($badname\)"\);   \($name\) = gethostbyname\($badname\);
&nbsp;  &Log\("alarm\(0\)"\);                  alarm\(0\);
\};
&Log\("done"\);

> /usr/ports/lang/perl5/work/perl\-5\.6\.1/perl ~/tmp/alarmtest\.pl
Sun Sep 15 12&#8203;:01&#8203;:53 PDT 2002 alarm\(10\)
Sun Sep 15 12&#8203;:01&#8203;:53 PDT 2002 gethostbyname\(197\.201\.144\.198\.orbs\.gst\-group\.co\.uk\)
Sun Sep 15 12&#8203;:02&#8203;:03 PDT 2002 done
\# see\, above 5\.6\.1 alarms in 10 seconds even
> /usr/local/bin/perl5\.8\.0 ~/tmp/alarmtest\.pl
Sun Sep 15 12&#8203;:02&#8203;:28 PDT 2002 alarm\(10\)
Sun Sep 15 12&#8203;:02&#8203;:28 PDT 2002 gethostbyname\(197\.201\.144\.198\.orbs\.gst\-group\.co\.uk\)
Sun Sep 15 12&#8203;:03&#8203;:49 PDT 2002 done
\# um\, above 5\.8\.0 alarms in 1 minute 21 seconds \-\- huh??

\-brian
briank@&#8203;briank\.com

Summary of my perl5 \(revision 5\.0 version 8 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=freebsd\, osvers=4\.6\.2\-release\, archname=i386\-freebsd
&nbsp;    uname='freebsd oe8\.briank\.com 4\.6\.2\-release freebsd 4\.6\.2\-release \#0&#8203;: thu aug 22 08&#8203;:49&#8203;:01 pdt 2002 root@&#8203;oe8\.briank\.com&#8203;:usrobjusrsrcsysold\_english\_800 i386 '
&nbsp;    config\_args=''
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-DHAS\_FPSETMASK \-DHAS\_FLOATINGPOINT\_H \-fno\-strict\-aliasing \-I/usr/local/include'\,
&nbsp;    optimize='\-O'\,
&nbsp;    cppflags='\-DHAS\_FPSETMASK \-DHAS\_FLOATINGPOINT\_H \-fno\-strict\-aliasing \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='2\.95\.3 20010315 \(release\) \[FreeBSD\]'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags ='\-Wl\,\-E  \-L/usr/local/lib'
&nbsp;    libpth=/usr/lib /usr/local/lib
&nbsp;    libs=\-lm \-lc \-lcrypt \-lutil
&nbsp;    perllibs=\-lm \-lc \-lcrypt \-lutil
&nbsp;    libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags='\-DPIC \-fpic'\, lddlflags='\-shared  \-L/usr/local/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: USE\_LARGE\_FILES
&nbsp;  Built under freebsd
&nbsp;  Compiled at Sep 11 2002 00&#8203;:42&#8203;:10
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /usr/local/lib/perl5/5\.8\.0/i386\-freebsd
&nbsp;    /usr/local/lib/perl5/5\.8\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.8\.0/i386\-freebsd
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.8\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.6\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.005
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    \.

**Comment by p5pRT** [comment]
### From @rgs

Brian Korver \(via RT\) wrote&#8203;:
> The following script demonstrates that the alarm\(\) call is broken
> in 5\.8\.0\, at least while running on FreeBSD 4\.6\.2&#8203;:
> 
> > cat alarmtest\.pl 
> $timeout = 10;
> \# use a host that will cause gethostbyname to wait for > $timeout seconds
> $badname = "197\.201\.144\.198\.orbs\.gst\-group\.co\.uk";
> 
> sub Log\(\) \{
>   $| = 1;
>   local\($d\) = \`date\`; chop\($d\);
>   print "$d @&#8203;\_\\n"
> \}
> 
> sub myALRM \{ die "alarm\\n" \}
> $SIG\{ALRM\} = 'myALRM';
> 
> eval \{
>   &Log\("alarm\($timeout\)"\);           alarm\($timeout\);
>   &Log\("gethostbyname\($badname\)"\);   \($name\) = gethostbyname\($badname\);

Well\, is there something in gethostbyname\(\) that may confuse alarm ?
The name resolver has some internal timeout value\, right ?

>   &Log\("alarm\(0\)"\);                  alarm\(0\);
> \};
> &Log\("done"\);
> 
> > /usr/ports/lang/perl5/work/perl\-5\.6\.1/perl ~/tmp/alarmtest\.pl
> Sun Sep 15 12&#8203;:01&#8203;:53 PDT 2002 alarm\(10\)
> Sun Sep 15 12&#8203;:01&#8203;:53 PDT 2002 gethostbyname\(197\.201\.144\.198\.orbs\.gst\-group\.co\.uk\)
> Sun Sep 15 12&#8203;:02&#8203;:03 PDT 2002 done
> \# see\, above 5\.6\.1 alarms in 10 seconds even
> > /usr/local/bin/perl5\.8\.0 ~/tmp/alarmtest\.pl
> Sun Sep 15 12&#8203;:02&#8203;:28 PDT 2002 alarm\(10\)
> Sun Sep 15 12&#8203;:02&#8203;:28 PDT 2002 gethostbyname\(197\.201\.144\.198\.orbs\.gst\-group\.co\.uk\)
> Sun Sep 15 12&#8203;:03&#8203;:49 PDT 2002 done
> \# um\, above 5\.8\.0 alarms in 1 minute 21 seconds \-\- huh??

**Comment by p5pRT** [comment]
### From briank@briank.com

"Rafael Garcia\-Suarez \(via RT\)" wrote&#8203;:
> 
> Brian Korver \(via RT\) wrote&#8203;:
> > The following script demonstrates that the alarm\(\) call is broken
> > in 5\.8\.0\, at least while running on FreeBSD 4\.6\.2&#8203;:
> >
> > > cat alarmtest\.pl
> > $timeout = 10;
> > \# use a host that will cause gethostbyname to wait for > $timeout seconds
> > $badname = "197\.201\.144\.198\.orbs\.gst\-group\.co\.uk";
> >
> > sub Log\(\) \{
> >   $| = 1;
> >   local\($d\) = \`date\`; chop\($d\);
> >   print "$d @&#8203;\_\\n"
> > \}
> >
> > sub myALRM \{ die "alarm\\n" \}
> > $SIG\{ALRM\} = 'myALRM';
> >
> > eval \{
> >   &Log\("alarm\($timeout\)"\);           alarm\($timeout\);
> >   &Log\("gethostbyname\($badname\)"\);   \($name\) = gethostbyname\($badname\);
> 
> Well\, is there something in gethostbyname\(\) that may confuse alarm ?
> The name resolver has some internal timeout value\, right ?

Why would the same script work fine under perl 5\.6\.1 but not
5\.8\.0?  See below\.\.\.\.


> 
> >   &Log\("alarm\(0\)"\);                  alarm\(0\);
> > \};
> > &Log\("done"\);
> >
> > > /usr/ports/lang/perl5/work/perl\-5\.6\.1/perl ~/tmp/alarmtest\.pl
> > Sun Sep 15 12&#8203;:01&#8203;:53 PDT 2002 alarm\(10\)
> > Sun Sep 15 12&#8203;:01&#8203;:53 PDT 2002 gethostbyname\(197\.201\.144\.198\.orbs\.gst\-group\.co\.uk\)
> > Sun Sep 15 12&#8203;:02&#8203;:03 PDT 2002 done
> > \# see\, above 5\.6\.1 alarms in 10 seconds even
> > > /usr/local/bin/perl5\.8\.0 ~/tmp/alarmtest\.pl
> > Sun Sep 15 12&#8203;:02&#8203;:28 PDT 2002 alarm\(10\)
> > Sun Sep 15 12&#8203;:02&#8203;:28 PDT 2002 gethostbyname\(197\.201\.144\.198\.orbs\.gst\-group\.co\.uk\)
> > Sun Sep 15 12&#8203;:03&#8203;:49 PDT 2002 done
> > \# um\, above 5\.8\.0 alarms in 1 minute 21 seconds \-\- huh??


\-\- 

\-brian
briank@&#8203;briank\.com

**Comment by p5pRT** [comment]
### From goldbb2@earthlink.net

Brian Korver \(via RT\) wrote&#8203;:
\[snip\]
> sub myALRM \{ die "alarm\\n" \}
> $SIG\{ALRM\} = 'myALRM';
> 
> eval \{
>   &Log\("alarm\($timeout\)"\);           alarm\($timeout\);
>   &Log\("gethostbyname\($badname\)"\);   \($name\) = gethostbyname\($badname\);
>   &Log\("alarm\(0\)"\);                  alarm\(0\);
> \};
> &Log\("done"\);

In perl 5\.8\.0\, signals are deferred until between opcodes\, and handlers
installed with the SA\_RESTART flag\.  As a result\, alarm\(\) will not
interrupt system calls\.  You must find some other way of timing out
gethostbyname\.

\-\- 
Announcement in the Zoo&#8203;:
"Please don't scare the ostriches\! Cement floor\!"

**Comment by p5pRT** [comment]
### From easmith@beatrice.rutgers.edu

On Sep 15\,  7&#8203;:43pm\, Benjamin Goldberg wrote&#8203;:
> Brian Korver \(via RT\) wrote&#8203;:
> \[snip\]
> > sub myALRM \{ die "alarm\\n" \}
> > $SIG\{ALRM\} = 'myALRM';
> > 
> > eval \{
> >   &Log\("alarm\($timeout\)"\);           alarm\($timeout\);
> >   &Log\("gethostbyname\($badname\)"\);   \($name\) = gethostbyname\($badname\);
> >   &Log\("alarm\(0\)"\);                  alarm\(0\);
> > \};
> > &Log\("done"\);
> 
> In perl 5\.8\.0\, signals are deferred until between opcodes\, and handlers
> installed with the SA\_RESTART flag\.

Why is SA\_RESTART being used\, then?

> As a result\, alarm\(\) will not interrupt system calls\.  You must find some
> other way of timing out gethostbyname\.

How?

&nbsp;    \-Allen

\-\- 
Allen Smith            http&#8203;://cesario\.rutgers\.edu/easmith/
September 11\, 2001        A Day That Shall Live In Infamy II
"They that can give up essential liberty to obtain a little temporary
safety deserve neither liberty nor safety\." \- Benjamin Franklin

**Comment by p5pRT** [comment]
### From easmith@beatrice.rutgers.edu

On Sep 15\,  8&#8203;:01pm\, Allen Smith wrote&#8203;:
> On Sep 15\,  7&#8203;:43pm\, Benjamin Goldberg wrote&#8203;:
> > Brian Korver \(via RT\) wrote&#8203;:
> > \[snip\]
> > > sub myALRM \{ die "alarm\\n" \}
> > > $SIG\{ALRM\} = 'myALRM';
> > > 
> > > eval \{
> > >   &Log\("alarm\($timeout\)"\);           alarm\($timeout\);
> > >   &Log\("gethostbyname\($badname\)"\);   \($name\) = gethostbyname\($badname\);
> > >   &Log\("alarm\(0\)"\);                  alarm\(0\);
> > > \};
> > > &Log\("done"\);
> > 
> > In perl 5\.8\.0\, signals are deferred until between opcodes\, and handlers
> > installed with the SA\_RESTART flag\.
> 
> Why is SA\_RESTART being used\, then?

>From an inspection of util\.c\, I determine that SA\_RESTART is \_not\_ being
used in the new model&#8203;:

\#ifdef SA\_RESTART
\#if defined\(PERL\_OLD\_SIGNALS\)
&nbsp;    act\.sa\_flags |= SA\_RESTART; /\* SVR4\, 4\.3\+BSD \*/
\#endif
\#endif

I suspect the problem is that the underlying gethostbyname library code is
seeing a EINTR from the networking functions and restarting instead of
properly exiting\.

&nbsp;    \-Allen

\-\- 
Allen Smith            http&#8203;://cesario\.rutgers\.edu/easmith/
September 11\, 2001        A Day That Shall Live In Infamy II
"They that can give up essential liberty to obtain a little temporary
safety deserve neither liberty nor safety\." \- Benjamin Franklin

**Comment by p5pRT** [comment]
### From briank@briank.com

"Allen Smith \(via RT\)" wrote&#8203;:
> From an inspection of util\.c\, I determine that SA\_RESTART is \_not\_ being
> used in the new model&#8203;:
> 
> \#ifdef SA\_RESTART
> \#if defined\(PERL\_OLD\_SIGNALS\)
>     act\.sa\_flags |= SA\_RESTART; /\* SVR4\, 4\.3\+BSD \*/
> \#endif
> \#endif
> 
> I suspect the problem is that the underlying gethostbyname library code is
> seeing a EINTR from the networking functions and restarting instead of
> properly exiting\.
> 
>         \-Allen

Why would that behavior have changed when I "upgraded" from
perl 4\.6\.1 to 5\.8\.0?

\-brian
briank@&#8203;briank\.com

**Comment by p5pRT** [comment]
### From nick.ing-simmons@elixent.com

Benjamin Goldberg \<goldbb2@&#8203;earthlink\.net> writes&#8203;:
>Brian Korver \(via RT\) wrote&#8203;:
>\[snip\]
>> sub myALRM \{ die "alarm\\n" \}
>> $SIG\{ALRM\} = 'myALRM';
>> 
>> eval \{
>>   &Log\("alarm\($timeout\)"\);           alarm\($timeout\);
>>   &Log\("gethostbyname\($badname\)"\);   \($name\) = gethostbyname\($badname\);
>>   &Log\("alarm\(0\)"\);                  alarm\(0\);
>> \};
>> &Log\("done"\);
>
>In perl 5\.8\.0\, signals are deferred until between opcodes\, 

True\.

>and handlers
>installed with the SA\_RESTART flag\.  

Not supposed to be true \- indeed another report recently complained
that this was no longer true\.

>As a result\, alarm\(\) will not
>interrupt system calls\.  You must find some other way of timing out
>gethostbyname\.

It think it is gethostbyname\(\) which is doing its own retry on EINTR\.

\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From @eserte

Brian Korver \(via RT\) \<perlbug@&#8203;perl\.org> writes&#8203;:

> \# New Ticket Created by  Brian Korver 
> \# Please include the string&#8203;:  \[perl \#17341\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt2/Ticket/Display\.html?id=17341 >
> 
> 
> The following script demonstrates that the alarm\(\) call is broken
> in 5\.8\.0\, at least while running on FreeBSD 4\.6\.2&#8203;:
> 
> > cat alarmtest\.pl 
> $timeout = 10;
> \# use a host that will cause gethostbyname to wait for > $timeout seconds
> $badname = "197\.201\.144\.198\.orbs\.gst\-group\.co\.uk";
> 
> sub Log\(\) \{
>   $| = 1;
>   local\($d\) = \`date\`; chop\($d\);
>   print "$d @&#8203;\_\\n"
> \}
> 
> sub myALRM \{ die "alarm\\n" \}
> $SIG\{ALRM\} = 'myALRM';
> 
> eval \{
>   &Log\("alarm\($timeout\)"\);           alarm\($timeout\);
>   &Log\("gethostbyname\($badname\)"\);   \($name\) = gethostbyname\($badname\);
>   &Log\("alarm\(0\)"\);                  alarm\(0\);
> \};
> &Log\("done"\);
> 
> > /usr/ports/lang/perl5/work/perl\-5\.6\.1/perl ~/tmp/alarmtest\.pl
> Sun Sep 15 12&#8203;:01&#8203;:53 PDT 2002 alarm\(10\)
> Sun Sep 15 12&#8203;:01&#8203;:53 PDT 2002 gethostbyname\(197\.201\.144\.198\.orbs\.gst\-group\.co\.uk\)
> Sun Sep 15 12&#8203;:02&#8203;:03 PDT 2002 done
> \# see\, above 5\.6\.1 alarms in 10 seconds even
> > /usr/local/bin/perl5\.8\.0 ~/tmp/alarmtest\.pl
> Sun Sep 15 12&#8203;:02&#8203;:28 PDT 2002 alarm\(10\)
> Sun Sep 15 12&#8203;:02&#8203;:28 PDT 2002 gethostbyname\(197\.201\.144\.198\.orbs\.gst\-group\.co\.uk\)
> Sun Sep 15 12&#8203;:03&#8203;:49 PDT 2002 done
> \# um\, above 5\.8\.0 alarms in 1 minute 21 seconds \-\- huh??
> 

Does it help if you use

&nbsp;    sigaction SIGALRM\, new POSIX&#8203;::SigAction sub \{ die "alarm\\n" \}
&nbsp;        or die "Error setting SIGALRM handler&#8203;: $\!\\n";

instead of setting $SIG\{ALRM\}? See perldoc perlvar for more
information\.

Regards\,
&nbsp;    Slaven

\-\- 
Slaven Rezic \- slaven\.rezic@&#8203;berlin\.de

&nbsp;    tkrevdiff \- graphical display of diffs between revisions \(RCS or CVS\)
&nbsp;    http&#8203;://ptktools\.sourceforge\.net/\#tkrevdiff

**Comment by p5pRT** [comment]
### From @lizmat

At 10&#8203;:15 AM 9/16/02 \+0200\, Slaven Rezic wrote&#8203;:
>Does it help if you use
>    sigaction SIGALRM\, new POSIX&#8203;::SigAction sub \{ die "alarm\\n" \}
>         or die "Error setting SIGALRM handler&#8203;: $\!\\n";
>instead of setting $SIG\{ALRM\}? See perldoc perlvar for more
>information\.

This fixes the problem under Linux\.  I used this demo program&#8203;: call with a
true parameter to use the standard $SIG\, else the sigaction route is taken\.

== signaltest ==============================================================
use POSIX '&#8203;:signal\_h';

if \(shift\) \{
&nbsp;     $SIG\{ALRM\} = sub \{die "alarm went off from \\$SIG\\n"\};
\} else \{
&nbsp;     sigaction SIGALRM\, new POSIX&#8203;::SigAction sub \{ die "alarm went off from 
sigaction\\n" \}
&nbsp;      or die "Error setting SIGALRM handler&#8203;: $\!\\n";
\}

alarm\( 1 \);
$domain = gethostbyaddr\( pack\( 'C4'\,split\(/\\\./\,'1\.2\.3\.4'\)\)\,2 \);
============================================================================

$ time perl signaltest
alarm went off from sigaction

real    0m1\.120s
user    0m0\.120s
sys     0m0\.010s

$ time perl signaltest 1
alarm went off from $SIG

real    0m10\.153s
user    0m0\.110s
sys     0m0\.010s

So\, only after the gethostbyaddr times out\, is the $SIG signal 
honoured\.  Whereas the sigaction way of doing things\, \_does\_ work\.

I guess I need to check on Thread&#8203;::Signal and Thread&#8203;::Pool&#8203;::Resolve\.\.\.  ;\-\)


Liz

**Comment by p5pRT** [comment]
### From easmith@beatrice.rutgers.edu

On Sep 16\,  8&#8203;:11am\, Elizabeth Mattijsen wrote&#8203;:
> At 10&#8203;:15 AM 9/16/02 \+0200\, Slaven Rezic wrote&#8203;:
> >Does it help if you use
> >    sigaction SIGALRM\, new POSIX&#8203;::SigAction sub \{ die "alarm\\n" \}
> >         or die "Error setting SIGALRM handler&#8203;: $\!\\n";
> >instead of setting $SIG\{ALRM\}? See perldoc perlvar for more
> >information\.
> 
> This fixes the problem under Linux\.  I used this demo program&#8203;: call with a
> true parameter to use the standard $SIG\, else the sigaction route is taken\.

\[\.\.\.\]

> So\, only after the gethostbyaddr times out\, is the $SIG signal 
> honoured\.  Whereas the sigaction way of doing things\, \_does\_ work\.

Good\, although this still leaves an unfortunate degree of incompatibility
with programs written using the original $SIG code \- not sure exactly what
to do about this at the moment\.

> I guess I need to check on Thread&#8203;::Signal and Thread&#8203;::Pool&#8203;::Resolve\.\.\.  ;\-\)

Hmm\.\.\. I note that\, unlike the %SIG stuff in util\.c\, there isn't any
checking on "which thread are we in?" in POSIX\.xs\. That sigprocmask\, which
should be thread\-specific in what signal is blocked\, is available is
helpful\, however\.

&nbsp;    \-Allen

\-\- 
Allen Smith            http&#8203;://cesario\.rutgers\.edu/easmith/
September 11\, 2001        A Day That Shall Live In Infamy II
"They that can give up essential liberty to obtain a little temporary
safety deserve neither liberty nor safety\." \- Benjamin Franklin

**Comment by p5pRT** [comment]
### From @jhi

I now added the POSIX&#8203;::sigaction\(\) trick as suggested by Slaven to perlipc\.pod\.
SInce I don't think there are other good ways around this \(except completely
retracting the safe signals\, which is unlikely\)\, I'm marking the problem ticket
as resolved\.

**Comment by p5pRT** [comment]
@jhi - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From guest@guest.guest.xxxxxxxx

On Sun Sep 15 12&#8203;:16&#8203;:55 2002\, briank@&#8203;briank\.com wrote&#8203;:
> The following script demonstrates that the alarm\(\) call is broken
> in 5\.8\.0\, at least while running on FreeBSD 4\.6\.2&#8203;:
> 
> > cat alarmtest\.pl
> $timeout = 10;
> \# use a host that will cause gethostbyname to wait for > $timeout
> seconds
> $badname = "197\.201\.144\.198\.orbs\.gst\-group\.co\.uk";
> 
> sub Log\(\) \{
>   $| = 1;
>   local\($d\) = \`date\`; chop\($d\);
>   print "$d @&#8203;\_\\n"
> \}
> 
> sub myALRM \{ die "alarm\\n" \}
> $SIG\{ALRM\} = 'myALRM';
> 
> eval \{
>   &Log\("alarm\($timeout\)"\);           alarm\($timeout\);
>   &Log\("gethostbyname\($badname\)"\);   \($name\) =
> gethostbyname\($badname\);
>   &Log\("alarm\(0\)"\);                  alarm\(0\);
> \};
> &Log\("done"\);
:

While this example \*may\* indicate that the SIGALRM behaviour is faulty 
in Perl 5\.8\.0\, it is nonetheless a very dangerous and non\-portable 
solution\.

The gethostbyname\(\) function is signal\-unsafe on most releases of 
FreeBSD that I have tested \(including 4\.x\, 5\.x\, and 6\.x\)\. You should 
\*never\* attempt to timeout gethostbyname\(\) using the eval/alarm/die 
construct\, otherwise you risk leaking KQUEUE file descriptors that may 
quickly max\-out the kernel's file descriptor table\.

If you require non\-blocking behavior of name\-server queries\, then 
Net&#8203;::DNS is by far the more stable alternative\.

\-\-Randall
